From 356510aad880ba09da42afa810ac98d234ca913a Mon Sep 17 00:00:00 2001
From: David Huff <dhuff@redhat.com>
Date: Thu, 10 Jul 2008 17:04:37 -0400
Subject: [PATCH] added get_partitions to kickstart.py

---
 imgcreate/kickstart.py |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/imgcreate/kickstart.py b/imgcreate/kickstart.py
index 0f797b8..0fbefb7 100644
--- a/imgcreate/kickstart.py
+++ b/imgcreate/kickstart.py
@@ -495,6 +495,9 @@ def get_groups(ks, required = []):
 def get_excluded(ks, required = []):
     return ks.handler.packages.excludedList + required
 
+def get_partitions(ks, required = []):
+    return ks.handler.partition.partitions
+
 def ignore_missing(ks):
     return ks.handler.packages.handleMissing == ksconstants.KS_MISSING_IGNORE
 
-- 
1.5.4.1

